### PR TITLE
fix(eslint-plugin): [consistent-type-imports] handle imports without specifiers

### LINKED ISSUE #7209: Bug: [consistent-type-imports] Attempts to fix the wrong import statement
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

@typescript-eslint/consistent-type-imports went wrong with side effect import and type import.

```
$ /workspaces/consistent-type-imports-repro/node_modules/.bin/eslint baz.ts

Oops! Something went wrong! :(

ESLint: 8.44.0

Error: Non-null Assertion Failed: Expected to find a } for the ImportDeclaration.
Occurred while linting /workspaces/consistent-type-imports-repro/baz.ts:1
Rule: "@typescript-eslint/consistent-type-imports"
    at Object.nullThrows (/workspaces/consistent-type-imports-repro/node_modules/@typescript-eslint/utils/dist/eslint-utils/nullThrows.js:23:15)
    at fixInsertNamedSpecifiersInNamedSpecifierList (/workspaces/consistent-type-imports-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/consistent-type-imports.js:406:44)
    at fixToValueImportDeclaration (/workspaces/consistent-type-imports-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/consistent-type-imports.js:614:55)
    at fixToValueImportDeclaration.next (<anonymous>)
    at Object.fix (/workspaces/consistent-type-imports-repro/node_modules/@typescript-eslint/eslint-plugin/dist/rules/consistent-type-imports.js:262:56)
    at fix.next (<anonymous>)
    at Function.from (<anonymous>)
    at normalizeFixes (/workspaces/consistent-type-imports-repro/node_modules/eslint/lib/linter/report-translator.js:182:33)
    at /workspaces/consistent-type-imports-repro/node_modules/eslint/lib/linter/report-translator.js:349:49
    at Object.report (/workspaces/consistent-type-imports-repro/node_modules/eslint/lib/linter/linter.js:1021:41)
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Reproduction Repository Link

https://github.com/OpportunityLiu/consistent-type-imports-repro

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn eslint baz.ts`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.0.0` |
| `@typescript-eslint/parser`            | `6.0.0` |
| `@typescript-eslint/scope-manager`     | `6.0.0` |
| `@typescript-eslint/typescript-estree` | `6.0.0` |
| `@typescript-eslint/type-utils`        | `6.0.0` |
| `@typescript-eslint/utils`             | `6.0.0` |
| `TypeScript`                           | `5.1.6` |
| `ESLint`                               | `8.44.0` |
| `node`                                 | `18.16.1` |

### PR DESCRIPTION
**Pull request by courageousillumination**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7209
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Currently side effect imports will be marked as a `valueOnlyNamedImport`, which causes an error when trying insert the correct value import because it is expecting a `}`. 

With these changes, we require that there is at least one specifier. If there is only a side effect import, a new import will be added with the specific value import. 

One note is that this will also impact a named import with no specifiers `import {} from 'foo';` Previously this would have inserted the value import into the empty `{}`; now it would insert it in a new import on a new line. I didn't see any way to distinguish between these two cases without looking for the braces explicitly; if that is not acceptable (since it changes how a fix is applied), please let me know and I can do the parsing.

This is my first PR against this repository, so please let me know if I missed anything 💖

### CODE REVIEW

File: packages/eslint-plugin/src/rules/consistent-type-imports.ts [hunk 1]
```diff
@@ -142,6 +142,7 @@ export default createRule<Options, MessageIds>({
               } else {
                 if (
                   !sourceImports.valueOnlyNamedImport &&
+                  node.specifiers.length &&
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/consistent-type-imports.ts:145]: > any way to distinguish between these two cases without looking for the braces explicitly

If you do end up needing to distinguish the two cases, `context` has some APIs around tokens.

```plaintext
> context.sourceCode.getText(node)
'import {} from 'foo';'

> context.getTokenBefore(node.source)
{type: 'Identifier', value: 'from', range: Array(2), loc: {…}, start: <accessor>, …}

> context.getTokens(node)
(6) [{…}, {…}, {…}, {…}, {…}, {…}]
0: {type: 'Keyword', value: 'import', range: Array(2), loc: {…}, start: <accessor>, …}
1: {type: 'Punctuator', value: '{', range: Array(2), loc: {…}, start: <accessor>, …}
2: {type: 'Punctuator', value: '}', range: Array(2), loc: {…}, start: <accessor>, …}
3: {type: 'Identifier', value: 'from', range: Array(2), loc: {…}, start: <accessor>, …}
4: {type: 'String', value: ''foo'', range: Array(2), loc: {…}, start: <accessor>, …}
5: {type: 'Punctuator', value: ';', range: Array(2), loc: {…}, start: <accessor>, …}
```

```plaintext
> context.sourceCode.getText(node)
'import 'foo';'

> context.getTokenBefore(node.source)
{type: 'Keyword', value: 'import', range: Array(2), loc: {…}, start: <accessor>, …}

> context.getTokens(node)
(3) [{…}, {…}, {…}]
0: {type: 'Keyword', value: 'import', range: Array(2), loc: {…}, start: <accessor>, …}
1: {type: 'String', value: ''foo'', range: Array(2), loc: {…}, start: <accessor>, …}
2: {type: 'Punctuator', value: ';', range: Array(2), loc: {…}, start: <accessor>, …}
```

Hope this helps!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
An excellent start for a first PR! Great 💖 emoji choice too, noticed & appreciated. 😄 

Requesting changes on adding a unit test (and rule logic if needed) for the extra case mentioned.

File: packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts
[the remark this answers is not in the document]
Comment (courageousillumination) [reply packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts:1]: Good call. I've added a test for this case. I've decided to stick with my original implementation; `import {} from 'foo'` seems abnormal enough that I'd like to maintain the authors intent (what that is, I'm not sure). That said, I can change up the code using some of the tools below if we want to go the other way.

File: packages/eslint-plugin/src/rules/consistent-type-imports.ts (hunk 1 above)
Comment (courageousillumination) [reply packages/eslint-plugin/src/rules/consistent-type-imports.ts:145]: Ah this is helpful. As outlined above, I've opted to keep the current implementation, but this is good context for future work.

File: packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts
[the remark this answers is not in the document]
Comment (ehoogeveen-medweb) [reply packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts:1]: IIRC `import {} from 'foo';` is semantically equivalent to `import 'foo';` - correct me if I'm wrong. So
```ts
import {
} from 'foo';
```
could be a diff-friendly way to reserve space for importing exported values from a module with side effects (ew). Mind you, [import/no-empty-named-blocks](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-empty-named-blocks.md) reports such blocks.

**Review by JoshuaKGoldberg — DISMISSED** [review]
LGTM, thanks! Very clean & straightforward. Nice. 🙂 

Adding `1 approval` and requesting review from @bradzacher because it coincidentally intersects with #8335.

File: packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts:1]: _Technically_ they're a little different. It's tricky 🥲. 

Looking just at TypeScript `target: CommonJS` as an example:
* `import {} from 'foo'` on its own doesn't emit a `require("foo")`
* Adding `import 'foo'` _does_ emit the `require("foo")`

#8335 has a whole deep dive into this fun area.

**Review by bradzacher — APPROVED** [review]

**Pull request merged**